### PR TITLE
create vaults and auctions release (proposal 76)

### PR DESCRIPTION
**Pull request by Chris-Hibbert**
closes: #171

## Description

add tests and assets for the [vaults and auctions release ](https://github.com/Agoric/agoric-sdk/releases/tag/agoric-upgrade-16av) **agoric-upgrade-16av**.

### CODE REVIEW

**Comment by dckc** [comment]
For reference, this PR represents the following mainnet governance decision:

 - [#76.  Replace Auctioneer and Upgrade Vaults](https://www.mintscan.io/agoric/proposals/76)
 Voting Time : Sep 1st 2024, 14:47:11 ~ Sep 4th 2024, 14:47:11

I presume it's ok to tweak the PR title to add `(proposal 76)`.

File: proposals/76:vaults-auctions/upgradeVaults.test.js [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+
+  t.true(newAuctioneerFromNewBundle(details));
+  // This query matches both the auction and its governor, so double the count
+  t.true(Object.keys(details).length > 2);
+};
+
+const verifyVaultPriceUpdate = async t => {
+  const quote = await getVaultPrices(0);
+
+  t.true(quote.value[0].amountIn.brand.includes(' ATOM '));
```
[the remark this answers is not in the document]
Comment (dckc) [reply proposals/76:vaults-auctions/upgradeVaults.test.js:105]: I checked. It's an alleged name. something like `$4.Alleged ATOM Brand`.

Options include:
 - let it slide. I've done too much of that lately.
 - add a comment telling other folks to "do as I say, not as I do" about reliability of alleged names
 - fix it by not throwing the boardIds that serve as clist identities.
 - ... others?

Let's discuss.

File: proposals/76:vaults-auctions/upgradeVaults.test.js [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+const oraclesByBrand = new Map();
+
+let roundId = 2;
+
+const setupOracles = async t => {
+  const atomOutPre = await getPriceQuote('ATOM');
+  t.is(atomOutPre, '+12010000');
+
+  await addPreexistingOracles('ATOM', oraclesByBrand);
```
Comment (dckc) [inline proposals/76:vaults-auctions/upgradeVaults.test.js:33]: `addPreexistingOracles` only uses the 1s arg to do `oraclesByBrand.set('ATOM', ..)`. Odd API. I'd expect the caller to do that.  I guess that pre-dates this PR, so IOU an issue on synthetic chain.

File: proposals/76:vaults-auctions/upgradeVaults.test.js [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+  getDetailsMatchingVats,
+} from './vatDetails.js';
+
+const oraclesByBrand = new Map();
+
+let roundId = 2;
+
+const setupOracles = async t => {
+  const atomOutPre = await getPriceQuote('ATOM');
+  t.is(atomOutPre, '+12010000');
```
Comment (dckc) [inline proposals/76:vaults-auctions/upgradeVaults.test.js:31]: `getPriceQuote` returns unparsed smallcaps, with object references disconnected from their boardIds. This tech debt is growing and growing. Here we compare what's actually an integer to its smallcaps string representation. At least that doesn't throw away info like object references.

I suppose it pre-dates this PR, so separate issue:
 - #176

File: proposals/76:vaults-auctions/upgradeVaults.test.js [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+  t.log('awaiting new quotes');
+  const atomOut = await getPriceQuote('ATOM');
+  t.is(atomOut, '+11200000');
+};
+
+export const BID_OFFER_ID = 'bid-vaultUpgrade-test3';
+const createNewBid = async t => {
+  await createBid('20', USER1ADDR, BID_OFFER_ID);
+  const liveOffer = await getLiveOffers(USER1ADDR);
+  t.true(liveOffer[0].includes(BID_OFFER_ID));
```
Comment (dckc) [inline proposals/76:vaults-auctions/upgradeVaults.test.js:49]:
```suggestion
  const liveOffers = await getLiveOffers(USER1ADDR);
  t.true(liveOffers[0].includes(BID_OFFER_ID));
```

and why check `.includes(BID_OFFER_ID)` rather than `=== BID_OFFER_ID`? Does `getLiveOffers` not return an array of offer ids?

File: proposals/76:vaults-auctions/upgradeVaults.test.js [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+  await createNewBid(t);
+
+  t.log('open a marginal vault');
+  await openMarginalVault(t);
+
+  t.log('trigger Auction');
+  await triggerAuction(t);
+
+  t.log('make new auction');
+  await checkAuctionVat(t);
```
Comment (dckc) [inline proposals/76:vaults-auctions/upgradeVaults.test.js:135]: Are we actually making a new auction vat here? this whole test runs _after_ the upgrade, yes? Here we might confirm that the auctioneer is new, but we're not making it, right?

File: proposals/76:vaults-auctions/vatDetails.js [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+      );
+    },
+    lookupVat,
+  });
+};
+
+// TODO(cth) get from synthetic-chain
+/** @param {string} vatName */
+export const getDetailsMatchingVats = async vatName => {
+  const fullPath = swingstorePath.replace(/^~/, NonNullish(HOME));
```
Comment (dckc) [inline proposals/76:vaults-auctions/vatDetails.js:80]: synthetic chain exports a `HOME` that might be null? strange.

Comment (dckc) [reply proposals/76:vaults-auctions/vatDetails.js:80]: elsewhere I see `env.HOME`. Why not use `env.HOME` here as well?

File: proposals/76:vaults-auctions/vatDetails.js [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+  const db = dbOpenAmbient(fullPath, { readonly: true });
```
Comment (dckc) [inline proposals/76:vaults-auctions/vatDetails.js:82]: The db should be passed into `getDetailsMatchingVats` per [OCap Discipline](https://github.com/Agoric/agoric-sdk/wiki/OCap-Discipline). Again, pre-dates this PR, so separate issue:

 - https://github.com/Agoric/agoric-sdk/issues/10038

**Review by dckc — COMMENTED** [review]
much of what I comment on pre-dates this PR, but let's please discuss relying on alleged names for correctness.

I didn't get all the way thru, but I've got a meeting coming up, so I'm sharing this much.

File: packages/synthetic-chain/package.json [hunk 8]
```diff
@@ -2,9 +2,7 @@
   "name": "@agoric/synthetic-chain",
   "version": "0.2.0",
   "description": "Utilities to build a chain and test proposals atop it",
-  "bin": {
-    "synthetic-chain": "dist/cli/cli.js"
-  },
+  "bin": "dist/cli/cli.js",
```
[the remark this answers is not in the document]
Comment (Chris-Hibbert) [reply packages/synthetic-chain/package.json:5]: AFAICT, this change was made by `yarn doctor`.  I don't know the consequences.

[the remark this answers is not in the document]
Comment (turadg) [reply packages/synthetic-chain/package.json:5]: it's something that Yarn and NPM disagree on. I'd suggest reverting this (to let NPM win) but not a blocker

File: proposals/76:vaults-auctions/upgradeVaults.test.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (turadg) [reply proposals/76:vaults-auctions/upgradeVaults.test.js:105]: Is your concern that the brand may have been forged somewhere in the history of A3P?

File: proposals/76:vaults-auctions/package.json [hunk 9]
```diff
@@ -0,0 +1,25 @@
+{
+  "agoricProposal": {
+    "type": "/agoric.swingset.CoreEvalProposal",
+    "releaseNotes": "https://github.com/Agoric/agoric-sdk/releases/tag/agoric-upgrade-16av",
+    "planName": "agoric-upgrade-16av"
```
[the remark this answers is not in the document]
Comment (turadg) [reply proposals/76:vaults-auctions/package.json:5]: please remove, it's confusing

File: proposals/76:vaults-auctions/upgradeVaults.test.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (dckc) [reply proposals/76:vaults-auctions/upgradeVaults.test.js:105]: My concern is that we're testing something that's not part of the API for brands.

[the remark this answers is not in the document]
Comment (turadg) [reply proposals/76:vaults-auctions/upgradeVaults.test.js:105]: Those would occur after this in history so can't affect this particular test.

File: proposals/76:vaults-auctions/README.md [hunk 10]
```diff
@@ -0,0 +1,4 @@
+# CoreEvalProposal to upgrade vaults and auctions
+
+The `assets` for this proposal were generated as part of [creating
+the release](https://github.com/Agoric/agoric-sdk/releases/tag/agoric-upgrade-16av).
```
[the remark this answers is not in the document]
Comment (Chris-Hibbert) [reply proposals/76:vaults-auctions/README.md]: done

File: proposals/76:vaults-auctions/package.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (Chris-Hibbert) [reply proposals/76:vaults-auctions/package.json]: done

File: proposals/76:vaults-auctions/saveAuctionInstance.js [hunk 11]
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env node
+import { recordAuctioneerInstance } from './vatDetails.js';
```
[the remark this answers is not in the document]
Comment (Chris-Hibbert) [reply proposals/76:vaults-auctions/saveAuctionInstance.js:2]: Moving some pieces to synthetic chain: https://github.com/Agoric/agoric-3-proposals/pull/174

File: proposals/76:vaults-auctions/submission/add-auction-permit.json [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+      "IST": true
+    }
+  },
+  "evaluateBundleCap": "makeCoreProposalBehavior",
+  "modules": {
+    "utils": {
+      "runModuleBehaviors": "makeCoreProposalBehavior"
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (Chris-Hibbert) [reply proposals/76:vaults-auctions/submission/add-auction-permit.json]: done

File: proposals/76:vaults-auctions/upgradeVaults.test.js (hunk 4 above)
Comment (Chris-Hibbert) [reply proposals/76:vaults-auctions/upgradeVaults.test.js]: I don't think it does. I think it's a larger structure I didn't know how to reliably untangle.

File: proposals/76:vaults-auctions/upgradeVaults.test.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (Chris-Hibbert) [reply proposals/76:vaults-auctions/upgradeVaults.test.js:107]: I'm okay with #2. A3P is a controlled environment in which there are no adversaries submitting faked names.

File: proposals/76:vaults-auctions/upgradeVaults.test.js (hunk 5 above)
Comment (Chris-Hibbert) [reply proposals/76:vaults-auctions/upgradeVaults.test.js:137]: The new auction vat already exists before the test starts.

File: proposals/76:vaults-auctions/vatDetails.js [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+
+export const recordAuctioneerInstance = async () => {
+  const oldAuctionInstance = await getInstanceBoardId('auctioneer');
+  assert(oldAuctionInstance, 'no auction instance found');
+  console.log('old auction instance ', oldAuctionInstance, env.HOME);
+
+  await writeFile(lastAuctionInstancePathname, oldAuctionInstance);
+};
+
+// duplicated from vat-status in synthetic-chain
```
[the remark this answers is not in the document]
Comment (Chris-Hibbert) [reply proposals/76:vaults-auctions/vatDetails.js:32]: There's a dance here because getting changes into synthetic-chain waits on a release cycle. I've been adding tools to each a3p test, then moving them to synthetic-chain to be shared once they've proved their worth.

Here I'm adding a method to the swingStore object that gets returned. It's being added to synthetic-chain in https://github.com/Agoric/agoric-3-proposals/pull/174, but it's not there yet.

File: proposals/76:vaults-auctions/vatDetails.js (hunk 6 above)
Comment (Chris-Hibbert) [reply proposals/76:vaults-auctions/vatDetails.js]: done

File: proposals/76:vaults-auctions/vatDetails.js (hunk 7 above)
Comment (Chris-Hibbert) [reply proposals/76:vaults-auctions/vatDetails.js:82]: This copy of `getDetailsMatchingVats` is here until there's a version published from `synthetic-chain`. 

Please suggest revisions in #174.

File: proposals/76:vaults-auctions/submission/upgrade-vaults-permit.json [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+      "auctioneer": true
+    }
+  },
+  "evaluateBundleCap": "makeCoreProposalBehavior",
+  "modules": {
+    "utils": {
+      "runModuleBehaviors": "makeCoreProposalBehavior"
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (Chris-Hibbert) [reply proposals/76:vaults-auctions/submission/upgrade-vaults-permit.json]: done

**Review by dckc — APPROVED** [review]
Good work

File: packages/synthetic-chain/public/upgrade-test-scripts/run_eval.sh [hunk 15]
```diff
@@ -6,7 +6,7 @@ set -eo pipefail
 
 source ./env_setup.sh
 
-PROPOSAL=$1
+export PROPOSAL=$1
```
Comment (turadg) [inline packages/synthetic-chain/public/upgrade-test-scripts/run_eval.sh:9]: please revert this change. It's unnecessary API surface to maintain.

The motivation I see is for the proposal to log what proposal it is, but it already knows.

**Review by turadg — CHANGES_REQUESTED** [review]

File: packages/synthetic-chain/public/upgrade-test-scripts/run_eval.sh (hunk 15 above)
Comment (Chris-Hibbert) [reply packages/synthetic-chain/public/upgrade-test-scripts/run_eval.sh:9]: without it, `run_prepare.sh` prints *[] Running prepare.sh*. with it, I get *[74:upgrade-15] Running prepare.sh*.

Comment (turadg) [reply packages/synthetic-chain/public/upgrade-test-scripts/run_eval.sh:9]: This `run_prepare.sh`? https://github.com/Agoric/agoric-3-proposals/blob/e7404a61a1b17b59081afce49fdaa02748be1e42/packages/synthetic-chain/public/upgrade-test-scripts/run_prepare.sh#L27-L43

That's supposed to exit if `PROPOSAL` is empty. 

Maybe you mean `proposals/76:vaults-auctions/eval.sh`? 

https://github.com/Agoric/agoric-3-proposals/blob/e7404a61a1b17b59081afce49fdaa02748be1e42/packages/synthetic-chain/public/upgrade-test-scripts/run_prepare.sh#L27-L43

That file is in `76:vaults-auctions` so it doesn't need the env to tell it what the proposal is

Comment (Chris-Hibbert) [reply packages/synthetic-chain/public/upgrade-test-scripts/run_eval.sh]: The code sure does say that. I had looked in my terminal window, and copied those progress fragments (**[] Running prepare.sh**  and **[74:upgrade-15] Running prepare.sh**) to show what caused me to add this line.

Now, I concede that all the places I see **Running prepare.sh** are preceded by the check for an empty string. I don't know what changed, but I'll drop this, even though it did resolve a problem I experienced.

**Review by turadg — APPROVED** [review]

File: proposals/76:vaults-auctions/package.json [hunk 16]
```diff
@@ -0,0 +1,24 @@
+{
+  "agoricProposal": {
+    "type": "/agoric.swingset.CoreEvalProposal",
+    "releaseNotes": "https://github.com/Agoric/agoric-sdk/releases/tag/agoric-upgrade-16av",
```
Comment (turadg) [inline proposals/76:vaults-auctions/package.json:4]: the build-proposals job is failing because this trailing comma is invalid JSON.

maybe we should add a "lint" job? it could report if `yarn doctor` reported or changed anything and also do eslint style linting on files.

File: proposals/76:vaults-auctions/upgradeVaults.test.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (Chris-Hibbert) [reply proposals/76:vaults-auctions/upgradeVaults.test.js:107]: We discussed this at the Zoe meeting. The issue here is more about comparing to strings extracted from vstorage without unmarshalling the capData. It seems straightforward for a3p tests to get unmarshalled data. @dckc  is investigating.
